sct.c

```diff
@@ -36,83 +40,115 @@ static void usage()
+        gammab += (crtc_gamma->blue[size - 1]);
 
-        if (size > 0)
+        XFree(crtc_gamma);
+    }
+    XFree(res);
+    gammam = (gammar > gammag) ? gammar : gammag;
+    gammam = (gammab > gammam) ? gammab : gammam;
+    if (gammam > 0.0)
+    {
+        gammar /= gammam;
+        gammag /= gammam;
+        gammab /= gammam;
+        gammad = gammab - gammar;
+        if (gammad < 0.0)
         {
-            double g_inv = (size + 1) / size / GAMMA_MULT;
-            gammar = crtc_gamma->red[size - 1] * g_inv;
-            gammag = crtc_gamma->green[size - 1] * g_inv;
-            gammab = crtc_gamma->blue[size - 1] * g_inv;
-            t = (64465 - 109049 * gammar + 46013 * gammar * gammar -
-                 4322 * gammag + 10708 * gammag * gammag -
-                 2662 * gammab + 1355 * gammab * gammab);
-            ts += t;
+            if (gammab == 0.0)
+            {
+                if (gammag == 0.0)
+                {
+                    t = TEMPERATURE_ZERO;
+                } else {
+                    t = gammag;
+                    t -= GAMMA_K0GR;
+                    t /= GAMMA_K1GR;
+                    t = exp(t);
+                    t += TEMPERATURE_ZERO;
+                }
+            } else {
+                t = gammag + gammad + 1.0;
+                t -= (GAMMA_K0GR + GAMMA_K0BR);
+                t /= (GAMMA_K1GR + GAMMA_K1BR);
+                t = exp(t);
+                t += TEMPERATURE_ZERO;
+            }
+        } else {
+            t = gammag + 1.0 - gammad;
+            t -= (GAMMA_K0GB + GAMMA_K0RB);
+            t /= (GAMMA_K1GB + GAMMA_K1RB);
+            t = exp(t);
+            t += (TEMPERATURE_NORM - TEMPERATURE_ZERO);
         }
-
-        XFree(crtc_gamma);
     }
-    temp = (int)(ts / (double)n);
 
-    XFree(res);
+    temp = (int)(t + 0.5);
+
     return temp;
 }
 
 static void sct_for_screen(Display *dpy, int screen, int temp)
 {
+    double t = 0.0, g = 0.0, gammar, gammag, gammab;
     Window root = RootWindow(dpy, screen);
     XRRScreenResources *res = XRRGetScreenResourcesCurrent(dpy, root);
 
-    temp -= TEMPERATURE_MIN;
-    double ratio = temp % TEMPERATURE_STEP / TEMPERATURE_STEP;
-    double gammar = AVG(r, temp, ratio);
-    double gammag = AVG(g, temp, ratio);
-    double gammab = AVG(b, temp, ratio);
+    t = (double)temp;
+    if (temp < TEMPERATURE_NORM)
+    {
+        gammar = 1.0;
+        if (temp < TEMPERATURE_ZERO)
+        {
+            gammag = 0.0;
+            gammab = 0.0;
+        } else {
+            g = log(t - TEMPERATURE_ZERO);
+            gammag = DoubleTrim(GAMMA_K0GR + GAMMA_K1GR * g, 0.0, 1.0);
+            gammab = DoubleTrim(GAMMA_K0BR + GAMMA_K1BR * g, 0.0, 1.0);         
+        }
+    } else {
+        g = log(t - (TEMPERATURE_NORM - TEMPERATURE_ZERO));
+        gammar = DoubleTrim(GAMMA_K0RB + GAMMA_K1RB * g, 0.0, 1.0);
+        gammag = DoubleTrim(GAMMA_K0GB + GAMMA_K1GB * g, 0.0, 1.0);
+        gammab = 1.0;
+        
```

Comment:
```suggestion
```
please remove this newline

Reply:
https://github.com/faf0/sct/pull/10/commits/4c2f7a1390a0fba58ea52632504122cb3a20bbd6